code/GraphQL/LinkablePlugin.php

```diff
@@ -52,43 +58,44 @@ public function apply(ModelQuery $query, Schema $schema, array $config = []): vo
         if ($class !== SiteTree::class && !is_subclass_of($class, SiteTree::class)) {
             return;
         }
-        Schema::invariant(
-            !$query->isList(),
-            'Plugin %s cannot be applied to queries that return lists. Query "%s" is a list',
-            static::getIdentifier(),
-            $query->getName()
+        $singleFieldName = $this->config()->get('field_name');
+        $pluraliser = $schema->getConfig()->getPluraliser();
+        $fieldName = $query->isList() ? $pluraliser($singleFieldName) : $singleFieldName;
+        $type = $query->isList() ? '[String]' : 'String';
+        $query->addArg($fieldName, $type);
+        $query->addResolverAfterware(
+            static::config()->get('resolver'),
+            ['fieldName' => $fieldName]
         );
-        $fieldName = $this->config()->get('field_name');
-        
-        $query->addArg($fieldName, 'String');
-        $query->addResolverAfterware([static::class, 'applyLinkFilter']);
     }
 
     /**
-     * @param $obj
-     * @param array $args
      * @param array $context
-     * @param ResolveInfo $info
-     * @param callable $done
-     * @return SiteTree|DataList|null
+     * @return callable
      */
-    public static function applyLinkFilter(
-        $obj,
-        array $args,
-        array $context,
-        ResolveInfo $info,
-        callable $done
-    ) {
-        $fieldName = static::config()->get('field_name');
-        $filterLink = $args['filter'][$fieldName] ?? null;
-        $argLink = $args[$fieldName] ?? null;
-        $filterLink = $filterLink ?: $argLink;
+    public static function applyLinkFilter(array $context): callable
```

Comment:
The function signature has totally changed, under semver this requires a new major release which we obviously cannot do.

Reply:
There is a 0.000000000000 chance that anyone is using this in the user space, but I've obeyed the semver gods and refactored this to be much uglier but backward compatible nonetheless. :)

It's probably worth discussing at some point that as soon as you add context to a resolver, it has to change its method signature, and that core resolvers should therefore be marked `@internal`. There's no good reason to be invoking those methods outside of a graphql request.